python/across_api/base/ephem.py

```diff
@@ -0,0 +1,161 @@
+# Copyright © 2023 United States Government as represented by the
+# Administrator of the National Aeronautics and Space Administration.
+# All Rights Reserved.
+
+from datetime import timedelta
+from typing import Optional
+
+import astropy.units as u  # type: ignore
+import numpy as np
+from astropy.constants import R_earth  # type: ignore
+from astropy.coordinates import (  # type: ignore
+    GCRS,
+    TEME,
+    CartesianDifferential,
+    CartesianRepresentation,
+    SkyCoord,
+    get_body,
+)
+from astropy.time import Time  # type: ignore
+from fastapi import HTTPException
+from sgp4.api import Satrec  # type: ignore
+
+from ..base.schema import TLEEntry
+from .common import round_time
+
+
+class EphemBase:
+    """
+    Base class for computing ephemeris data, for spacecraft whose positions can
+    be determined using a Two-Line Element (TLE), i.e. most Earth-orbiting
+    spacecraft. This ephemeris primarily calculates the position of the
+    spacecraft in GCRS coordinates. It also calculates the position of the Sun
+    and Moon in GCRS coordinates, and the latitude and longitude of the
+    spacecraft over the Earth.
+
+    The design of this ephemeris class is based upon how actual LEO spacecraft
+    ephemeris calculations work, and therefore is designed to mimic constraints
+    calculated onboard a spacecraft, to as closely match and predict those as
+    possible.
+
+    Parameters
+    ----------
+    begin
+        The start time of the ephemeris.
+    end
+        The end time of the ephemeris.
+    stepsize
+        The time step size, the default is 60 seconds.
+
+    Attributes
+    ----------
+    earth_radius
+        The radius of the Earth in degrees. If not specified (None), it will be
+        calculated based on the distance from the Earth to the spacecraft.
+    tle
+        The TLE (Two-Line Element) data for the satellite.
+
+    Methods
+    -------
+    ephindex
+        Returns the array index for a given time.
+    """
+
+    # Type hints
+    begin: Time
+    end: Time
+    stepsize: u.Quantity
+    earth_radius: Optional[u.Quantity] = None
+    tle: Optional[TLEEntry] = None
+
+    def __len__(self) -> int:
+        return len(self.timestamp)
+
+    def ephindex(self, t: Time) -> int:
+        """
+        For a given time, return an index for the nearest time in the
+        ephemeris. Note that internally converting from Time to datetime makes
+        this run way faster.
+
+        Parameters
+        ----------
+        t
+            The time to find the nearest index for.
+
+        Returns
+        -------
+            The index of the nearest time in the ephemeris.
+        """
+        return int(np.argmin(np.abs((self.timestamp.datetime - t.datetime))))
+
+    def __init__(self, begin: Time, end: Time, stepsize: u.Quantity = 60 * u.s):
```

Comment:
Is this where we should use our `DateRangeDep` and `StepSizeDep` definitions that are currently in `api.py`?

Reply:
As above, those are only used in API endpoints, not in the ephem code.... and yes, `DateRangeDep` is another orphaned Depend from the removed Ephemeris API endpoint, so I removed that also in c68918e.